src/view/com/post-thread/PostThreadItem.tsx

```diff
@@ -661,28 +673,68 @@ function ExpandedPostDetails({
   record,
   needsTranslation,
   translatorUrl,
+  setTranslatedText,
 }: {
   post: AppBskyFeedDefs.PostView
   record?: AppBskyFeedPost.Record
   needsTranslation: boolean
   translatorUrl: string
+  setTranslatedText: React.Dispatch<React.SetStateAction<string>>
 }) {
   const pal = usePalette('default')
   const {_} = useLingui()
   const openLink = useOpenLink()
+  const langPrefs = useLanguagePrefs()
 
   const text = record?.text || ''
 
-  const onTranslatePress = React.useCallback(() => {
+  const [didTranslate, setDidTranslate] = useState<boolean>(false)
+  const [isTranslating, setIsTranslating] = useState<boolean>(false)
+
+  const onTranslatePress = React.useCallback(async () => {
+    if (isTranslating) return
+
     if (
       isNativeTranslationAvailable &&
       isLanguageSupported(record?.langs?.at(0))
     ) {
-      NativeTranslationModule.presentAsync(text)
+      if (isIOS) {
+        NativeTranslationModule.presentAsync(text)
+      } else if (isAndroid) {
+        if (didTranslate) {
+          setTranslatedText(text)
+          setDidTranslate(false)
+        } else {
+          try {
+            setIsTranslating(true)
+            const translated = await NativeTranslationModule.translateAsync(
+              getPostLanguage(post) ?? 'en',
```

Comment:
worth noting that MLKit also has a "language identification" API that could be used to get the post language, but I figured its better to use the ATProto's API here